### PR TITLE
Server connects to its listen socket addr

### LINKED ISSUE #46800: torch in cloud(randomly sche) may have server connect timeout problem
TcpStore listen at all addresses (0.0.0.0), as it listen at `port`

https://github.com/pytorch/pytorch/blob/789e9353044eb2227462ef470e4b362ac025be4d/torch/lib/c10d/TCPStore.cpp#L301-L310

a pytorch server proc will start a tcp-daemon and then connect to it by `address:port`

recently, we have this case, launch the pytorch in two nodes (n1:192.x.x.0, n2:192.x.x.1), and due to the proc of pytorch is randomly scheduled to a node by the scheduler, so we can't simply assume the server proc is running at n1 or n2

to clarify

n1: the worker proc is running, but server_addr is n1
n2: the server proc is running, but server_addr is n1

server proc and worker proc will timeout after 300s left

though this problem can be solved by the node index, as we can force the server proc running at the index 1 of the node rather then other node

but i propose to add some protected codes or logs around L301 - L310

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @xush6528 @osalpekar @jiayisuse @agolynski

### PR DESCRIPTION
**Pull request by zrss**
Fixes #46800

### CODE REVIEW

File: torch/lib/c10d/TCPStore.cpp [hunk 1]
```diff
@@ -301,9 +301,16 @@ TCPStore::TCPStore(
   if (isServer_) {
     // Opening up the listening socket
     std::tie(masterListenSocket_, tcpStorePort_) = tcputil::listen(masterPort);
+    
+    std::string socketAddr = tcputil::getSocketAddr(masterListenSocket_);
+    LOG(INFO) << "master listen socket addr: " << socketAddr;
```
Comment (pritamdamania87) [inline torch/lib/c10d/TCPStore.cpp:306]: Nit: "Master" instead of "master".

File: torch/lib/c10d/Utils.cpp [hunk 2]
```diff
@@ -344,16 +352,12 @@ std::tuple<int, std::string> accept(
...
-  socklen_t addrLen = sizeof(addr);
-  SYSCHECK_ERR_RETURN_NEG1(::getpeername(
-      socket, reinterpret_cast<struct ::sockaddr*>(&addr), &addrLen))
-
   setSocketNoDelay(socket);
 
-  return std::make_tuple(
-      socket, sockaddrToString(reinterpret_cast<struct ::sockaddr*>(&addr)));
+  // Get address of the connecting process
+  std::string socketAddr = getSocketAddr(socket);
```
Comment (pritamdamania87) [inline torch/lib/c10d/Utils.cpp:358]: Just call `getSocketAddr(socket)` in make_tuple to avoid the extra variable.

**Review by pritamdamania87 — CHANGES_REQUESTED** [review]
Thanks for putting up a PR to fix this issue!

Can we add tests in TCPStoreTest.cpp to test this functionality? Also, it looks like there are a bunch of CI failures, could you fix those as well?

File: torch/lib/c10d/TCPStore.cpp (hunk 1 above)
Comment (zrss) [reply torch/lib/c10d/TCPStore.cpp]: done

File: torch/lib/c10d/Utils.cpp (hunk 2 above)
Comment (zrss) [reply torch/lib/c10d/Utils.cpp]: ok

**Comment by zrss** [comment]
@pritamdamania87 thanks for taking a look ~ i have checked the test code in `TCPStoreTest.cpp` (bootstrap server tcpstore, then mock clients connect and set key/value into the server store), currently, from my point of view it can cover the new added codes of mine. PTAL. or could u give me some advices about testing as i am a freshman in pytorch :)

**Comment by pritamdamania87** [comment]
@zrss Looks like there are a bunch of CI failures related to TCPStoreTest, could you look into those and fix them?

**Comment by zrss** [comment]
@pandeykartikey , yeah, will dive into it

**Comment by zrss** [comment]
@pritamdamania87@rohan-varma, PTAL

File: torch/lib/c10d/TCPStore.cpp [hunk 3]
```diff
@@ -301,9 +301,16 @@ TCPStore::TCPStore(
   if (isServer_) {
     // Opening up the listening socket
     std::tie(masterListenSocket_, tcpStorePort_) = tcputil::listen(masterPort);
+    
+    std::string socketAddr = tcputil::getLocalSocketAddr(masterListenSocket_);
+    LOG(INFO) << "Master listen socket addr: " << socketAddr;
```
Comment (rohan-varma) [inline torch/lib/c10d/TCPStore.cpp:306]: Can we demote this to a `LOG(DEBUG)`? It might be too verbose to have this show up in INFO logs.

**Review by rohan-varma — COMMENTED** [review]
Thanks for this change! As @pritamdamania87 suggested, it would be good to have some unittests to confirm the desired behavior in `TCPStoreTest.cpp`.

**Comment by zrss** [comment]
@rohan-varma , get it ~ i'd like to add a case to make sure the wrong master addr won't break the master bootstrap

**Comment by zrss** [comment]
@rohan-varma , add a test func, PTAL

**Comment by zrss** [comment]
@rohan-varma @pritamdamania87 , hello, is there anything new about this topic, or do we plan further optimize the api of TCPStore, maybe taking the `TcpDaemon` out from the TCPStore ?

**Review by pritamdamania87 — APPROVED** [review]
LGTM

**Comment by malfet** [comment]
Reverted the change as it introduced a deadlock in Windows tests, for example see https://app.circleci.com/pipelines/github/pytorch/pytorch/246018/workflows/e031ac54-7455-4243-8056-b3e2e40df7f9/jobs/9345786

**Comment by zrss** [comment]
sorry for a broken, tcpstore won't work in win32 when i do these changes, currently it seems we add the support of win32, i should take sometime to fix it

**Pull request closed without merging**